sqlx-core/src/postgres/listen.rs

```diff
@@ -0,0 +1,188 @@
+        let cmd = PgConnection::build_listen_query(channel);
+        let _ = conn.execute(cmd.as_str(), Default::default()).await?;
+        Ok(PgListener::new(conn))
+    }
+
+    /// Fetch a new connection from the pool and register it as a listener on the specified channels.
+    pub async fn listen_all(
+        &self,
+        channels: impl IntoIterator<Item = impl AsRef<str>>,
+    ) -> Result<PgListener<PgPoolConnection>> {
+        let mut conn = self.acquire().await?;
+        for channel in channels {
+            let cmd = PgConnection::build_listen_query(&channel);
+            let _ = conn.execute(cmd.as_str(), Default::default()).await?;
+        }
+        Ok(PgListener::new(conn))
+    }
+}
+
+impl PgPoolConnection {
+    /// Fetch a new connection from the pool and register it as a listener on the specified channel.
+    pub async fn listen(mut self, channel: &impl AsRef<str>) -> Result<PgListener<Self>> {
+        let cmd = PgConnection::build_listen_query(channel);
+        let _ = self.execute(cmd.as_str(), Default::default()).await?;
+        Ok(PgListener::new(self))
+    }
+
+    /// Fetch a new connection from the pool and register it as a listener on the specified channels.
+    pub async fn listen_all(
+        mut self,
+        channels: impl IntoIterator<Item = impl AsRef<str>>,
+    ) -> Result<PgListener<Self>> {
+        for channel in channels {
+            let cmd = PgConnection::build_listen_query(&channel);
+            let _ = self.execute(cmd.as_str(), Default::default()).await?;
+        }
+        Ok(PgListener::new(self))
+    }
+}
+
+/// A stream of async database notifications.
+///
+/// Notifications will always correspond to the channel(s) specified this object is created.
+pub struct PgListener<C>(C);
+
+impl<C> PgListener<C> {
+    /// Construct a new instance.
+    pub(self) fn new(conn: C) -> Self {
+        Self(conn)
+    }
+}
+
+impl<C> PgListener<C>
+where
+    C: AsMut<PgConnection>,
+{
+    /// Get the next async notification from the database.
+    pub async fn next(&mut self) -> Result<NotifyMessage> {
+        loop {
+            match self.0.as_mut().receive().await? {
+                Some(Message::NotificationResponse(notification)) => return Ok(notification.into()),
+                // TODO: verify with team if this is correct. Looks like the connection being closed will cause an error
+                // to propagate up from `recevie`, but it would be good to verify with team.
+                Some(_) | None => continue,
+            }
+        }
+    }
+}
+
+impl<C> PgListener<C>
+where
+    C: Connection,
+{
+    /// Close this listener stream and its underlying connection.
+    pub async fn close(self) -> BoxFuture<'static, Result<()>> {
+        self.0.close()
+    }
+}
+
+impl<C> std::ops::Deref for PgListener<C> {
+    type Target = C;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<C> std::ops::DerefMut for PgListener<C> {
```

Comment:
If we're going to allow use of `PgConnection` with active subscriptions, we need to patch either `.receive()` or usages of it to push `NotificationResponse` messages to a `Vec` because they might be interleaved with responses from other commands. And then `PgListener.next()` should check that vec first before calling `.receive()`.

Reply:
Yea, I was think along those lines as well. On one hand, if the connection is in active use, then perhaps we don't worry about them, and we only handle notifications received while actually waiting for notifications.

On the other hand, we could create a buffer of the events as you mentioned. I would be hesitant to create an unlimited buffer of events because that could bloat the vector pretty heavily depending on the use case. I definitely see async notifications as being a bit transient as well. Some users may consider buffering of the events to be an anti-pattern, where they only want to see events as they occur. So, leaving it up to the user may be best.

What do you think about creating a configuration parameter in the `listen/listen_all` methods which will allow the user to configure message buffering? Perhaps an enum like:

```rust
enum BufferStrategy {
    All,
    Max(u32),
    None,
}
```

Then we just update the `PgConnection` state to embed this as an optional configuration (only `Some(_)` if `listen/listen_all` was used on the connection) — say, `buffer_strategy: Option<BufferStrategy>` — and then we buffer using the given strategy.

The only real complication with this strategy is that we will have to update the signature of `PgConnection.receive` to take some sort of parameter — say, `is_listening: bool` (or whatever) — so that it knows not to buffer when it is being called from the `PgListen.next` method.

The algorithm for that subsection of receive would be:
- if a NotificationMessage is received on the wire and `buffer_strategy.is_none()` then drop the notification.
- if a NotificationMessage is received on the wire and `buffer_strategy.is_some()` and `is_listening`, then immediately return the notification.
- if a NotificationMessage is received on the wire and `buffer_strategy.is_some()` and `!is_listening`, then buffer the notification using the configured strategy.

With that, we would just need to update `PgListener.next` to consult the notifications buffer before attempting to call `receive`. That may work pretty well. Leaving it up to the user is always nice too.

@abonander @mehcode thoughts?

Reply:
I'm leaning toward agreeing with this approach, although what about a `.buffer_limit(limit: impl Into<Option<usize>>)` method on `PgListener` instead? `None` would be "don't buffer" and the user would have to explicitly choose to make the buffer (effectively) unbounded with `usize::max_value()`.

Are we also thinking the buffer should be a FIFO queue with the oldest messages being popped first when the limit is reached? That way the user can set a limit of 1 and effectively have a "latest message only" subscription, although for that case I think supporting "latest message per topic" is strictly more useful (but can be implemented later).

Reply:
@mehcode looked through `libpq` to see what they do, and it looks like they just push messages to an unbounded linked list: https://github.com/postgres/postgres/blob/7559d8ebfa11d98728e816f6b655582ce41150f3/src/interfaces/libpq/fe-protocol3.c#L1415

So we can already do them one better by using a more datastructure for real computer architecture (`Vec`/`VecDeque`).

Reply:
Yea, double ended queue is definitely the way to go here, drop items off of the front as newer items come in which would breach the capacity.

Adding the `.buffer_limit(..)` method is a good idea. API would be something like:
- if not specified, the default will be an unbounded deque.
- if `None` is provided, then buffering will be disabled.
- if `Some(usize)` is provided, then we limit the capacity of the buffer to the specified size.

Sound about right? As far as the logging ... maybe we just expose a method so that the user can check the buffer size themselves, and if they want to log that information, we allow them to do so. Something like `PgListener.buffer_size() -> Option<usize>` (option because buffering may be disabled). Thoughts?

Reply:
Ok, this is effectively resolved now per: https://github.com/launchbadge/sqlx/pull/98#issuecomment-580370236